File: services/price-service/internal/adapters/base/xml.go
package base

import (
	"regexp"
	"strings"

	"github.com/kosarica/price-service/internal/adapters/config"
	"github.com/kosarica/price-service/internal/parsers/xml"
	"github.com/kosarica/price-service/internal/types"
)

// XmlAdapterConfig contains configuration for XML-based chain adapters
type XmlAdapterConfig struct {
	BaseAdapterConfig
	FieldMapping           xml.XmlFieldMapping
	AlternativeFieldMapping *xml.XmlFieldMapping
	DefaultItemsPath       string
	ItemPaths              []string
}

// BaseXmlAdapter provides common XML parsing logic
type BaseXmlAdapter struct {
	*BaseChainAdapter
	fieldMapping    xml.XmlFieldMapping
	altMapping      *xml.XmlFieldMapping
	itemPaths       []string
}

// NewBaseXmlAdapter creates a new base XML adapter
func NewBaseXmlAdapter(cfg XmlAdapterConfig) (*BaseXmlAdapter, error) {
	// Set XML file extension pattern
	if cfg.FileExtensionPattern == nil {
		cfg.FileExtensionPattern = regexp.MustCompile(`\.(xml|XML)$`)
	}

	// Create base adapter
	base, err := NewBaseChainAdapter(cfg.BaseAdapterConfig)
	if err != nil {
		return nil, err
	}

	// Set default item paths if not provided
	itemPaths := cfg.ItemPaths
	if len(itemPaths) == 0 {
		itemPaths = []string{
			"products.product",
			"Products.Product",
			"items.item",
			"Items.Item",
			"data.product",
			"Data.Product",
			"Cjenik.Proizvod",
			"cjenik.proizvod",
		}
	}

	return &BaseXmlAdapter{
		BaseChainAdapter: base,
		fieldMapping:     cfg.FieldMapping,
		altMapping:       cfg.AlternativeFieldMapping,
		itemPaths:        itemPaths,
	}, nil
}

// Parse parses XML content into normalized rows
// Tries multiple item paths and field mappings to find valid data
func (a *BaseXmlAdapter) Parse(content []byte, filename string, options *types.ParseOptions) (*types.ParseResult, error) {
	// Extract store identifier from filename
	storeIdentifier := a.extractStoreIdentifierFromFilename(filename)

	var lastErr error

	// Try with primary field mapping first
	for _, itemsPath := range a.itemPaths {
		result, err := a.parseWithItemsPath(content, itemsPath, a.fieldMapping, storeIdentifier)
		if err != nil {
			lastErr = err
			continue
		}
		if result.ValidRows > 0 {
			return result, nil
		}
	}

	// Try alternative field mapping if available
	if a.altMapping != nil {
		for _, itemsPath := range a.itemPaths {
			result, err := a.parseWithItemsPath(content, itemsPath, *a.altMapping, storeIdentifier)
			if err != nil {
				lastErr = err
				continue
			}
			if result.ValidRows > 0 {
				return result, nil
			}
		}
	}

	// Return error if all attempts failed
	if lastErr != nil {
		return nil, lastErr
	}

	// Return empty result as fallback
	return &types.ParseResult{
		Rows:      []types.NormalizedRow{},
		Errors:    []types.ParseError{},
		Warnings:  []types.ParseWarning{},
		TotalRows: 0,
		ValidRows: 0,
	}, nil
}

// parseWithItemsPath attempts to parse XML with specific items path and field mapping
func (a *BaseXmlAdapter) parseWithItemsPath(content []byte, itemsPath string, fieldMapping xml.XmlFieldMapping, storeIdentifier string) (*types.ParseResult, error) {
	// XML parsing will be fully implemented in Phase 7
	// For now, return a not-implemented error to avoid silent success
	return nil, &AdapterError{
		Chain: a.name,
		Msg:   "XML parsing not yet implemented - see Phase 7",
	}
}

// ExtractStoreIdentifier extracts store identifier from XML file
// For XML files, store ID is typically embedded in content or filename
func (a *BaseXmlAdapter) ExtractStoreIdentifier(file types.DiscoveredFile) *types.StoreIdentifier {
	// Try to extract from metadata if set during discovery
	if storeID, ok := file.Metadata["storeId"]; ok && storeID != "" {
		return &types.StoreIdentifier{
			Type:  "portal_id",
			Value: storeID,
		}
	}

	// Try to extract from filename as fallback
	identifier := a.extractStoreIdentifierFromFilename(file.Filename)
	if identifier != "" {
		return &types.StoreIdentifier{
			Type:  "filename_code",
			Value: identifier,
		}
	}

	return nil
}

// extractStoreIdentifierFromFilename extracts store identifier from XML filename
// Handles XML-specific patterns like "store_123" or "poslovnica_456"
func (a *BaseXmlAdapter) extractStoreIdentifierFromFilename(filename string) string {
	baseName := a.fileExtensionPattern.ReplaceAllString(filename, "")

	cleanName := baseName
	for _, pattern := range a.filenamePrefixPatterns {
		cleanName = pattern.ReplaceAllString(cleanName, "")
	}
	cleanName = strings.TrimSpace(cleanName)

	// Try to extract store ID from patterns like "store_123" or "poslovnica_456"
	storeIdPatterns := []string{
		`(?:store|poslovnica|trgovina)[_-]?(\d+)`,
		`(?:store|poslovnica|trgovina)[_-]?([A-Za-z0-9]+)`,
	}

	for _, pattern := range storeIdPatterns {
		re := regexp.MustCompile(pattern)
		if match := re.FindStringSubmatch(cleanName); len(match) > 1 {
			return match[1]
		}
	}

	if cleanName == "" {
		return a.fileExtensionPattern.ReplaceAllString(filename, "")
	}

	return cleanName
}

// GetFieldMapping returns the primary field mapping
func (a *BaseXmlAdapter) GetFieldMapping() xml.XmlFieldMapping {
	return a.fieldMapping
}

// GetAlternativeMapping returns the alternative field mapping
func (a *BaseXmlAdapter) GetAlternativeMapping() *xml.XmlFieldMapping {
	return a.altMapping
}

// GetItemPaths returns the item paths to try
func (a *BaseXmlAdapter) GetItemPaths() []string {
	return a.itemPaths
}

File: src/ingestion/adapters/base/xml.ts
import { type XmlFieldMapping, XmlParser } from "../../parsers/xml";
import type {
	DiscoveredFile,
	ParseOptions,
	ParseResult,
	StoreIdentifier,
} from "../../types";
import { type BaseAdapterConfig, BaseChainAdapter } from "./chain";

export interface XmlAdapterConfig {
	baseConfig: BaseAdapterConfig;
	fieldMapping: XmlFieldMapping;
	alternativeFieldMapping?: XmlFieldMapping;
	defaultItemsPath?: string;
	itemPaths?: string[];
}

export class BaseXmlAdapter extends BaseChainAdapter {
	protected fieldMapping: XmlFieldMapping;
	protected altMapping?: XmlFieldMapping;
	protected itemPaths: string[];

	constructor(cfg: XmlAdapterConfig) {
		super({
			...cfg.baseConfig,
			fileExtensionPattern:
				cfg.baseConfig.fileExtensionPattern ?? /\.(xml|XML)$/,
		});

		this.fieldMapping = cfg.fieldMapping;
		this.altMapping = cfg.alternativeFieldMapping;
		this.itemPaths = cfg.itemPaths ?? [
			"products.product",
			"Products.Product",
			"items.item",
			"Items.Item",
			"data.product",
			"Data.Product",
			"Cjenik.Proizvod",
			"cjenik.proizvod",
		];

		if (
			cfg.defaultItemsPath &&
			!this.itemPaths.includes(cfg.defaultItemsPath)
		) {
			this.itemPaths.unshift(cfg.defaultItemsPath);
		}
	}

	async parse(
		content: Buffer,
		filename: string,
		_options?: ParseOptions,
	): Promise<ParseResult> {
		const storeIdentifier = this.extractStoreIdentifierFromFilename(filename);
		let lastResult: ParseResult | null = null;

		for (const itemsPath of this.itemPaths) {
			const parser = new XmlParser({
				itemsPath,
				fieldMapping: this.fieldMapping,
				defaultStoreIdentifier: storeIdentifier,
				attributePrefix: "@_",
				encoding: "auto",
			});
			const result = parser.parseWithItemsPath(
				content,
				itemsPath,
				this.fieldMapping,
				storeIdentifier,
			);
			lastResult = result;
			if (result.validRows > 0) {
				return result;
			}
		}

		if (this.altMapping) {
			for (const itemsPath of this.itemPaths) {
				const parser = new XmlParser({
					itemsPath,
					fieldMapping: this.altMapping,
					defaultStoreIdentifier: storeIdentifier,
					attributePrefix: "@_",
					encoding: "auto",
				});
				const result = parser.parseWithItemsPath(
					content,
					itemsPath,
					this.altMapping,
					storeIdentifier,
				);
				lastResult = result;
				if (result.validRows > 0) {
					return result;
				}
			}
		}

		return (
			lastResult ?? {
				rows: [],
				errors: [],
				warnings: [],
				totalRows: 0,
				validRows: 0,
			}
		);
	}

	extractStoreIdentifier(file: DiscoveredFile): StoreIdentifier | null {
		const storeId = file.metadata?.storeId;
		if (storeId) {
			return { type: "portal_id", value: storeId };
		}
		const fallback = this.extractStoreIdentifierFromFilename(file.filename);
		if (fallback) {
			return { type: "filename_code", value: fallback };
		}
		return null;
	}

	protected extractStoreIdentifierFromFilename(filename: string): string {
		const baseName = filename.replace(this.fileExtensionPattern, "");
		let cleanName = baseName;
		for (const pattern of this.filenamePrefixPatterns) {
			cleanName = cleanName.replace(pattern, "");
		}
		cleanName = cleanName.trim();

		const patterns = [
			/(?:store|poslovnica|trgovina)[_-]?(\d+)/i,
			/(?:store|poslovnica|trgovina)[_-]?([A-Za-z0-9]+)/i,
		];
		for (const pattern of patterns) {
			const match = cleanName.match(pattern);
			if (match?.[1]) {
				return match[1];
			}
		}

		if (!cleanName) {
			return filename.replace(this.fileExtensionPattern, "");
		}
		return cleanName;
	}
}
